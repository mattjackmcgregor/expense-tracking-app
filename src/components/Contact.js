import React from 'react'

const Contact = () => (
  <div>this is the contact page</div>
)

export default Contact