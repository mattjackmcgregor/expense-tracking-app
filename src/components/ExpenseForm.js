import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'
// import 'react-dates/initialize'
// import 'react-dates/lib/css/_datepicker.css'

const mom = moment().format('MMM Do YYYY')
console.log(mom)


export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? props.expense.amount / 100 : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
    this.onDescriptionChange=this.onDescriptionChange.bind(this)
    this.onNoteChange=this.onNoteChange.bind(this)
    this.onAmountChange=this.onAmountChange.bind(this)
    this.onDateChange=this.onDateChange.bind(this)
    this.onFocusChange=this.onFocusChange.bind(this)
    this.submit=this.submit.bind(this)
  }
  onDescriptionChange(e) {
    const description = e.target.value  //setting to description varilabe for easy destructuring below
    this.setState(() => ({description})) 
  }
  onNoteChange(e) {
    const note = e.target.value 
    this.setState(() => ({note}))
  }
  onAmountChange(e) {
    const amount = e.target.value
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      this.setState(() =>({amount}))
    }
  }
  onDateChange(createdAt) {
    if(createdAt) {
      this.setState(() => ({createdAt}))
    }
  }
  
  onFocusChange({focused}) {
    this.setState(() =>({calendarFocused: focused}))
  }

  submit(e) {
    e.preventDefault()
    if(!this.state.description && !this.state.amount) {
      //error message
      this.setState(() =>({error: 'Please enter a description and amount'}))
    } else if(!this.state.description){
      this.setState(() =>({error: 'Please enter a description' }))
    } else if (!this.state.amount) {
      this.setState(() =>({error: 'Please enter an amount' }))
    } else {
      //submit
      this.setState(() =>({error: ''}))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) *100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }
  render() {
    return (
      <form className='form' onSubmit={this.submit}> 
      {this.state.error && <p className='form__error'>{this.state.error}</p>}
        <input 
          className='text-input'
          type='text' 
          placeholder='Description'
          value={this.state.description}
          autoFocus
          onChange={this.onDescriptionChange}
        />
        <input
          className='text-input'
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        
        <textarea
          className='input-textarea'
          type='text'
          placeholder='add a note for your expense (optional)'
          value={this.state.note}
          onChange={this.onNoteChange}
        />
        <div>
        <button className='button'>Save Expense</button>
        </div>
      </form>
    )
  }
}