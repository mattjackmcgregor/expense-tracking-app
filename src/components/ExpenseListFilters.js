import React from 'react'
import { connect } from 'react-redux';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters'
import {DateRangePicker} from 'react-dates'
import 'react-dates/initialize'
import 'react-dates/lib/css/_datepicker.css'


export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      calendarFocused: null
    }
    this.onDatesChange=this.onDatesChange.bind(this)
    this.onFocusChange=this.onFocusChange.bind(this
      )
  }
  onDatesChange({startDate, endDate}) {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }
  onFocusChange(calendarFocused) {
    this.setState(() => ({calendarFocused}))
  }
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate()
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount()
    }
  }
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }
  render() {
    return (
      <div className='content-container'>
        <div className="input-group">
          <div className='input-group__item'>
            <input className='text-input' type='text' placeholder='Search Expenses' value={this.props.filters.text} onChange={this.onTextChange} />
          </div>
          <div className='input-group__item'>
            <select className='input-select' value={this.props.filters.sortBy} onChange={this.onSortChange}>
              <option value="date">Date</option>
              <option value='amount'>Amount</option>
            </select>
          </div>
          <div className='input-group__item'>
            <DateRangePicker
              startDate={this.props.filters.startDate}
              startDateId='start date'
              endDate={this.props.filters.endDate}
              endDateId='end date'
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    )
  }
}
// const ExpenseListFilters = (props) => (
//   <div>
//     <input type='text' value={props.filters.text} onChange={(e) => {
//       props.dispatch(setTextFilter(e.target.value))
//     }} />
//     <select value={props.filters.sortBy} onChange={(e) => {
//       if (e.target.value === 'date') {
//         props.dispatch(sortByDate())
//       } else if (e.target.value === 'amount') {
//         props.dispatch(sortByAmount())
//       }
//     }}>
//       <option value="date">Date</option>
//       <option value='amount'>Amount</option>
//     </select>
//   </div>
// )

const mapStateToProps = (state) => (
  {
    filters: state.filters
  }
)
 

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setTextFilter: (text) => dispatch(setTextFilter(text))

})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)