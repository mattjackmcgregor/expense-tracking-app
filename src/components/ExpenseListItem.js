import React from 'react'
import {Link} from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'


// const ExpenseListItem = (props) => (
//   <div>
//     <p>description: {props.description}</p>
//     <p>Amount: {props.amount}</p>
//     <p>Created at: {props.createdAt}</p>
//   </div>
// )

//betterway descructuring
const ExpenseListItem = ({id, description, amount, createdAt}) => (
  
  <Link className='list-item' to={`/edit/${id}`}>
    <div>
      <h4 className='list-item__title'>{description}</h4>
      <span className='list-item__date'>{moment(createdAt).format('MMMM Do YYYY')}</span>
    </div>
    <div>
      <h3 className='list-item__amount'>{numeral(amount/100).format('$0,0.00')}</h3>
    </div>
  </Link>

    
  
)

export default (ExpenseListItem)