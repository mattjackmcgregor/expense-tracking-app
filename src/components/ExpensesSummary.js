import React from 'react'
import {connect} from 'react-redux'
import getVisibleExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/expenses-total'
import numeral from 'numeral'


export const ExpensesSummary = ({expensesCount, expensesTotal}) => {
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses'
  const formattedExpenseTotal = numeral(expensesTotal / 100).format('$0,00.00')
  return (
    <div>
      <h3>Viewing {expensesCount} {expenseWord} totaling {formattedExpenseTotal} </h3>
    </div>
  )
}
  
 



const mapStateToProps = (state) => {
  const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
  return {
    expensesCount: visibleExpenses.length,
    expensesTotal: getExpensesTotal(visibleExpenses)
  }
}
export default connect(mapStateToProps)(ExpensesSummary)