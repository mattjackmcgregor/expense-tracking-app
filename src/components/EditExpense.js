import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpense extends React.Component {
  editExpense = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense)
    console.log('updated', expense)
    this.props.history.push('/dashboard')
  }
  removeExpense = (id) => {
    this.props.startRemoveExpense({id: this.props.expense.id})
    this.props.history.push('/dashboard')
  }
  render() {
    return (
    <div>
      <div className='page-header'>
        <div className='content-container'>
          <h1 className='page-header__title'>Edit Expense</h1>
        </div>
      </div>
      <div className='content-container'>
        <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.editExpense}
        />
        <button className='button--secondary' onClick={this.removeExpense}>Remove Expense</button>
      </div>
    </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}
const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpense)