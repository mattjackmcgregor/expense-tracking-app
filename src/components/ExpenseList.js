import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'


export const ExpenseList = (props) => (
  <div className='content-container'>
    {
      props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        props.expenses.map((expense) => (
          // <ExpenseListItem 
          //   key={expense.id}
          //   description={expense.description}
          //   amount={expense.amount}
          //   createdAt={expense.createdAt}
          // />
    
          //betterway using spread opperator to destruct it
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )
    }
    
    
  </div>
)

const mapStateToProps = (state) => { //connect the stores state to the component props
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList) //this is essentially a HOC just xporting it straight instead of setting it to a variable 