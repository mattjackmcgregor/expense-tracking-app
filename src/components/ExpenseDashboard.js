import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboard = () => (
  <div>
    this is the ExpenseDashboard
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)

export default ExpenseDashboard