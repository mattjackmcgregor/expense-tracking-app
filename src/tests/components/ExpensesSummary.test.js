import React from 'react'
import {shallow} from 'enzyme'
import {ExpensesSummary} from '../../components/ExpensesSummary'

test('should correctly render ExpensesSummary with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={300}  />)
  expect(wrapper).toMatchSnapshot()
})
test('should correctly render ExpensesSummary with multiple expense', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={3} expensesTotal={1200}  />)
  expect(wrapper).toMatchSnapshot()
})