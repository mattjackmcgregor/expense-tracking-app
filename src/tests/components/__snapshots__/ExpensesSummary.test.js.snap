// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h3
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totaling 
      <span>
        $3.00
      </span>
    </h3>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h3
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totaling 
      <span>
        $12.00
      </span>
    </h3>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
